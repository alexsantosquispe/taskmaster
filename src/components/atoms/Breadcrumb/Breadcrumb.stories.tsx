import { Wrapper } from '@/utils/testing/unitTest.util';
import type { Meta, StoryObj } from '@storybook/react';
import { Breadcrumb } from './Breadcrumb';

const BREADCRUMBS = [
  {
    id: 'projects',
    label: 'Projects',
    href: ''
  },
  {
    id: 'proj-001',
    label: 'TaskMaster Mobile App',
    href: ''
  },
  {
    id: 'task-1',
    label: 'Task-001',
    href: ''
  },
  {
    id: 'sub-task-5',
    label: 'SubTask-005',
    href: ''
  }
];

const meta = {
  title: 'Atoms/Breadcrumb',
  component: Breadcrumb,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Breadcrumb>;

export default meta;

type Story = StoryObj<typeof Breadcrumb>;

export const Default: Story = () => {
  return (
    <Wrapper>
      <Breadcrumb items={BREADCRUMBS} />
    </Wrapper>
  );
};

Default.args = {};
