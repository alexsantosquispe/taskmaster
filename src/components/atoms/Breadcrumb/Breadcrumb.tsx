import cn from 'clsx';
import { twMerge } from 'tailwind-merge';
import { ChevronUpIcon } from '../../../icons';

type BreadcrumbItem = {
  id: string;
  label: string;
  href: string;
};

interface BreadcrumbProps {
  items: BreadcrumbItem[];
  className?: string;
}

export const Breadcrumb = ({ items, className }: BreadcrumbProps) => {
  return (
    <div
      className={twMerge(
        'text-primary/80 flex items-center gap-2 text-[0.8125rem] dark:text-white/70',
        className
      )}
    >
      {items.map((item, index) => {
        const isLastItem = index === items.length - 1;
        return (
          <div key={item.id} className="flex items-center gap-2 font-medium">
            <a
              href={item.href}
              className={twMerge(
                'hover:cursor-pointer',
                cn({ 'text-primary font-semibold dark:text-white': isLastItem })
              )}
            >
              <span>{item.label}</span>
            </a>
            {!isLastItem && <ChevronUpIcon className="size-3 rotate-90" />}
          </div>
        );
      })}
    </div>
  );
};
