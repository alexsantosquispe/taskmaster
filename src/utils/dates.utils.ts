import { differenceInSeconds, formatDistanceToNowStrict } from 'date-fns';

export const formatDateToDayMonth = (date: string) => {
  const options: Intl.DateTimeFormatOptions = {
    day: 'numeric',
    month: 'short'
  };
  return new Intl.DateTimeFormat('en-US', options).format(new Date(date));
};

export const formatToFriendlyDate = (date: string): string => {
  const parsedDate = new Date(date);

  if (isNaN(parsedDate.getTime())) return 'Invalid date';

  const diffSeconds = differenceInSeconds(new Date(), parsedDate);

  if (diffSeconds < 60) return 'just now';

  return `${formatDistanceToNowStrict(parsedDate)} ago`;
};
