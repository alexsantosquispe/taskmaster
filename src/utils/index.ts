import clsx from 'clsx';
import { PRIORITY_COLORS } from '../constants';
import { type PriorityType } from '../models/types';

export const formatDateToDayMonth = (date: string) => {
  const options: Intl.DateTimeFormatOptions = {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  };
  return new Intl.DateTimeFormat('en-US', options).format(new Date(date));
};

export const getPercentageStyles = (percentage: number) => {
  return clsx({
    'stroke-green-600 dark:stroke-green-500 text-green-600 dark:text-green-500':
      percentage >= 70,
    'stroke-green-500 dark:stroke-green-400 text-green-500 dark:text-green-400':
      percentage >= 56 && percentage < 70,
    'stroke-amber-500 dark:stroke-amber-400 text-amber-500 dark:text-amber-400':
      percentage >= 1 && percentage < 56,
    'stroke-primary dark:stroke-white text-neutral-600 dark:text-white/80':
      percentage === 0
  });
};

export const getPriorityStyles = (priority: PriorityType) => {
  return PRIORITY_COLORS[priority];
};
