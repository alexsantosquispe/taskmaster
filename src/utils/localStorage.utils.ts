import type { LSValuesTypes } from '../models/types';

export const getLSValue = (key: string) => localStorage.getItem(key);

export const setLSValue = (key: string, value: LSValuesTypes) => {
  localStorage.setItem(key, String(value));
};
