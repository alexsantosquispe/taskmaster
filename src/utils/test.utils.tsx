import { ToastStack } from '@/components/atoms/ToastStack/ToastStack';
import { ThemeProvider } from '@/contexts/ThemeProvider';
import { ToastProvider } from '@/contexts/ToastProvider';
import { apiClient } from '@/services/api';
import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { type ReactElement, type ReactNode } from 'react';
import { useForm, type Control, type FieldValues } from 'react-hook-form';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { twMerge } from 'tailwind-merge';

interface RenderWithFormOptions {
  defaultValues: FieldValues;
  componentToRender: (control: Control<FieldValues>) => ReactElement;
}

export const renderWithForm = ({
  defaultValues,
  componentToRender
}: RenderWithFormOptions) => {
  const Wrapper = () => {
    const { control } = useForm<FieldValues>({ defaultValues });
    return componentToRender(control);
  };

  return render(<Wrapper />);
};

interface WrapperProps {
  children: ReactNode;
  className?: string;
}

export const WrapperUI = ({ children, className }: WrapperProps) => {
  return (
    <div
      className={twMerge(
        'text-primary flex items-center justify-center rounded border border-neutral-400 p-8 dark:text-white',
        className
      )}
    >
      {children}
    </div>
  );
};

export const TestWrapper = ({ children }: WrapperProps) => {
  return (
    <ThemeProvider>
      <BrowserRouter
        future={{
          v7_startTransition: true,
          v7_relativeSplatPath: true
        }}
      >
        {children}
      </BrowserRouter>
    </ThemeProvider>
  );
};

export const ToastProviderWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <ToastProvider>
      <ToastStack />
      {children}
    </ToastProvider>
  );
};

export const ReduxWrapper = ({ children }: { children: ReactNode }) => {
  const store = configureStore({
    reducer: {
      [apiClient.reducerPath]: apiClient.reducer
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiClient.middleware)
  });

  return <Provider store={store}>{children}</Provider>;
};
