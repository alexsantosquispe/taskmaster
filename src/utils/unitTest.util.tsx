import { BrowserRouter } from 'react-router-dom';
import type { ReactNode } from 'react';
import { ThemeProvider } from '../contexts/ThemeProvider';

export const TestWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <ThemeProvider>
      <BrowserRouter
        future={{
          v7_startTransition: true,
          v7_relativeSplatPath: true
        }}
      >
        {children}
      </BrowserRouter>
    </ThemeProvider>
  );
};
