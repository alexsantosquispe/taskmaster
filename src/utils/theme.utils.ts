import { THEME_KEY } from '../constants';
import { Theme, type ThemeType } from '../contexts/ThemeContext';
import { getLSValue } from './localStorage.utils';

export const getSystemTheme = () => {
  if (typeof window === 'undefined') return Theme.light;

  return window.matchMedia('(prefers-color-scheme: dark)').matches
    ? Theme.dark
    : Theme.light;
};

export const getStoredTheme = (): ThemeType | null => {
  const storedTheme = getLSValue(THEME_KEY);

  if (!storedTheme) return null;

  return storedTheme as ThemeType;
};

export const applyThemeToDocument = (theme: ThemeType) => {
  const root = window.document.documentElement;

  if (theme === Theme.dark) {
    root.classList.add(Theme.dark);
  } else {
    root.classList.remove(Theme.dark);
  }
};
