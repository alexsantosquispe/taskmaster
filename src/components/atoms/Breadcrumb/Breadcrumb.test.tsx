import { TestWrapper } from '@/utils/test.utils';
import { render } from '@testing-library/react';
import { Breadcrumb } from './Breadcrumb';

const BREADCRUMBS = [
  {
    id: 'projects',
    label: 'Projects',
    href: ''
  },
  {
    id: 'taskmaster-mobile-app',
    label: 'TaskMaster Mobile App',
    href: ''
  }
];

describe('Breadcrumb', () => {
  it('should render the component correctly', () => {
    const component = render(
      <TestWrapper>
        <Breadcrumb items={BREADCRUMBS} />
      </TestWrapper>
    );

    expect(component).toMatchSnapshot();
  });
});
