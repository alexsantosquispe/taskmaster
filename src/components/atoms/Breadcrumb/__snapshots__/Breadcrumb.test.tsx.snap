// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Breadcrumb should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="text-primary/80 flex items-center gap-2 text-[0.8125rem] dark:text-white/70"
      >
        <div
          class="flex items-center gap-2 font-medium"
        >
          <a
            class="hover:cursor-pointer"
            href="/"
          >
            <span>
              Projects
            </span>
          </a>
          <svg
            class="size-3 rotate-90"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m18 15-6-6-6 6"
            />
          </svg>
        </div>
        <div
          class="flex items-center gap-2 font-medium"
        >
          <a
            class="hover:cursor-pointer text-primary font-semibold dark:text-white"
            href="/"
          >
            <span>
              TaskMaster Mobile App
            </span>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="text-primary/80 flex items-center gap-2 text-[0.8125rem] dark:text-white/70"
    >
      <div
        class="flex items-center gap-2 font-medium"
      >
        <a
          class="hover:cursor-pointer"
          href="/"
        >
          <span>
            Projects
          </span>
        </a>
        <svg
          class="size-3 rotate-90"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m18 15-6-6-6 6"
          />
        </svg>
      </div>
      <div
        class="flex items-center gap-2 font-medium"
      >
        <a
          class="hover:cursor-pointer text-primary font-semibold dark:text-white"
          href="/"
        >
          <span>
            TaskMaster Mobile App
          </span>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
